src/main/java/com/uid2/admin/vertx/service/AdminKeyService.java

```diff
@@ -73,122 +77,127 @@ public AdminKeyService(JsonObject config,
 
         router.get("/api/admin/reveal").handler(
-                auth.handle(this::handleAdminReveal, Role.ADMINISTRATOR));
+                auth.handle(audit.handle(this::handleAdminReveal), Role.ADMINISTRATOR));
 
-        router.post("/api/admin/add").blockingHandler(auth.handle((ctx) -> {
+        router.post("/api/admin/add").blockingHandler(auth.handle(audit.handle((ctx) -> {
             synchronized (writeLock) {
-                this.handleAdminAdd(ctx);
+                return this.handleAdminAdd(ctx);
             }
-        }, Role.ADMINISTRATOR));
+        }), Role.ADMINISTRATOR));
 
-        router.post("/api/admin/del").blockingHandler(auth.handle((ctx) -> {
+        router.post("/api/admin/del").blockingHandler(auth.handle(audit.handle((ctx) -> {
             synchronized (writeLock) {
-                this.handleAdminDel(ctx);
+                return this.handleAdminDel(ctx);
             }
-        }, Role.ADMINISTRATOR));
+        }), Role.ADMINISTRATOR));
 
-        router.post("/api/admin/disable").blockingHandler(auth.handle(ctx -> {
+        router.post("/api/admin/disable").blockingHandler(auth.handle(audit.handle(ctx -> {
             synchronized (writeLock) {
-                this.handleAdminDisable(ctx);
+                return this.handleAdminDisable(ctx);
             }
-        }, Role.ADMINISTRATOR));
+        }), Role.ADMINISTRATOR));
 
-        router.post("/api/admin/enable").blockingHandler(auth.handle(ctx -> {
+        router.post("/api/admin/enable").blockingHandler(auth.handle(audit.handle(ctx -> {
             synchronized (writeLock) {
-                this.handleAdminEnable(ctx);
+                return this.handleAdminEnable(ctx);
             }
-        }, Role.ADMINISTRATOR));
+        }), Role.ADMINISTRATOR));
 
-        router.post("/api/admin/rekey").blockingHandler(auth.handle(ctx -> {
+        router.post("/api/admin/rekey").blockingHandler(auth.handle(audit.handle(ctx -> {
             synchronized (writeLock) {
-                this.handleAdminRekey(ctx);
+                return this.handleAdminRekey(ctx);
             }
-        }, Role.ADMINISTRATOR));
+        }), Role.ADMINISTRATOR));
 
-        router.post("/api/admin/roles").blockingHandler(auth.handle(ctx -> {
+        router.post("/api/admin/roles").blockingHandler(auth.handle(audit.handle(ctx -> {
             synchronized (writeLock) {
-                this.handleAdminRoles(ctx);
+                return this.handleAdminRoles(ctx);
             }
-        }, Role.ADMINISTRATOR));
+        }), Role.ADMINISTRATOR));
     }
 
-    private void handleAdminMetadata(RoutingContext rc) {
+    private OperationModel handleAdminMetadata(RoutingContext rc) {
         try {
             rc.response()
                     .putHeader(HttpHeaders.CONTENT_TYPE, "application/json")
                     .end(adminUserProvider.getMetadata().encode());
+            return new OperationModel(Type.ADMIN, null, Actions.LIST,
```

Comment:
A hardcoded key seems better. On the other hand, I don't think getting metadata is worth auditing. Same with admin list below

Reply:
I'll add a hardcoded key